nrf52-hal-common/src/uarte.rs

```diff
@@ -0,0 +1,153 @@
+pub struct Uarte<T>(T);
+
+impl<T> Uarte<T>
+where
+    T: UarteExt,
+{
+    pub fn new(uarte: T, mut pins: Pins, parity: Parity, baudrate: Baudrate) -> Self {
+        // Select pins
+        pins.rxd.set_high();
+        uarte.psel.rxd.write(|w| {
+            let w = unsafe { w.pin().bits(pins.rxd.pin) };
+            w.connect().connected()
+        });
+        pins.txd.set_high();
+        uarte.psel.txd.write(|w| {
+            let w = unsafe { w.pin().bits(pins.txd.pin) };
+            w.connect().connected()
+        });
+
+        // Optional pins
+        uarte.psel.cts.write(|w| {
+            if let Some(ref pin) = pins.cts {
+                let w = unsafe { w.pin().bits(pin.pin) };
+                w.connect().connected()
+            } else {
+                w.connect().disconnected()
+            }
+        });
+
+        uarte.psel.rts.write(|w| {
+            if let Some(ref pin) = pins.rts {
+                let w = unsafe { w.pin().bits(pin.pin) };
+                w.connect().connected()
+            } else {
+                w.connect().disconnected()
+            }
+        });
+
+        // Enable UARTE instance
+        uarte.enable.write(|w| w.enable().enabled());
+
+        // Configure
+        uarte.config.write(|w| {
+            w.hwfc()
+                .bit(pins.rts.is_some() && pins.cts.is_some())
+                .parity()
+                .variant(parity)
+        });
+
+        // Configure frequency
+        uarte.baudrate.write(|w| w.baudrate().variant(baudrate));
+
+        Uarte(uarte)
+    }
+
+    /// Write via UART
+    ///
+    /// This method uses transmits all bytes in `tx_buffer`
+    ///
+    /// The buffer must have a length of at most 255 bytes
+    pub fn write(&mut self, tx_buffer: &[u8]) -> Result<(), Error> {
+        // This is overly restrictive. See (similar SPIM issue):
+        // https://github.com/nrf-rs/nrf52/issues/17
+        if tx_buffer.len() > u8::max_value() as usize {
+            return Err(Error::TxBufferTooLong);
+        }
+
+        // Set up the DMA write
+        self.0.txd.ptr.write(|w|
+            // We're giving the register a pointer to the stack. Since we're
+            // waiting for the UART transaction to end before this stack pointer
+            // becomes invalid, there's nothing wrong here.
+            //
+            // The PTR field is a full 32 bits wide and accepts the full range
+            // of values.
+            unsafe { w.ptr().bits(tx_buffer.as_ptr() as u32) });
+        self.0.txd.maxcnt.write(|w|
+            // We're giving it the length of the buffer, so no danger of
+            // accessing invalid memory. We have verified that the length of the
+            // buffer fits in an `u8`, so the cast to `u8` is also fine.
+            //
+            // The MAXCNT field is 8 bits wide and accepts the full range of
+            // values.
+            unsafe { w.maxcnt().bits(tx_buffer.len() as _) });
```

Comment:
I think this is fine as it is, but we should make a mental note (or better yet, open an issue) about creating a DMA API, so all the different peripherals don't have to duplicate this code.

Reply:
It's probably good to think about. Right now I am working on a more actively DMA driven interface, which allows the user to trade the `Uarte` struct for basically a `(Sender<&[u8]>, Receiver<&[u8]>)` tuple, backed by a set of `heapless` `RingBuffer`s.

This would effectively double-buffer the incoming and outgoing messages, and the user can always `free()` the tuple pair to get a `UARTE0` or `Uarte` object back.

Once it actually works, I'll chime back in. IMO, we can have multiple different DMA abstractions (e.g. as a blocking tx/rx, as an async stream, etc.) that can coexist in the code, and the `UARTE0` periph is just used as an ownership token forcing only one to be active at a time.

Reply:
I was thinking about something less ambitious: Just a `write` method that takes a slice and handles the `ptr` and `maxcnt` registers (plus all the associated casting and `unsafe`) under the hood. Basically, just a friendlier interface that the HAL implementations can use internally.

Thinking about higher-level abstractions along the lines you presented here is beneficial too, of course. I don't have much of an opinion about how that should look like yet.

Reply:
Pushed my WIP here as `uarte0-async` in case you feel like looking (totally not ready for serious review, doesn't work at all yet): https://github.com/nrf-rs/nrf52-hal/blob/uarte0-async/nrf52-hal-common/src/uarte_stream.rs

The big things to do still:
* Implement the interrupt handler (basically check which event triggered, reload and restart the buffer if there is something left to do)
* Write interfaces for the `Handle` structs, basically shove stuff in the buffer and trigger an event if not already in progress, probably by setting a flag and calling the inner handler directly
* Figure out if there is a less insane way of setting the handler-dispatch, allowing for different inner functions to service the interrupt, depending on mode.
* Figure out the interface for configuring buffer length/passing in buffers (mostly around lifetimes). I might do something along the lines of [this](https://github.com/jamesmunns/mcf-config-demo)
* Figure out what to do about early-flushing of the receive buffer. Right now, there isn't really a way to flush out the items from the DMA without polling. One of the forums mentioned tying a GPIO (internally?) to the RX line, and using a timer to check for idle states longer than some window, or we could use a 1-shot time to check for the amount of data received (and if unchanged for some amount of time), but I don't have a great elegant answer here.

Sorry to go off track of this review :)